File: rust/2024/t01.rs
mod util;
use std::collections::HashMap;

fn read_pairs(file_name: &str) -> Vec<(i32, i32)> {
    util::read_lines(file_name)
        .into_iter()
        .map(|line| {
            let parts: Vec<i32> = line.split("   ")
                .map(|part| part.parse::<i32>().unwrap())
                .collect();
            (parts[0], parts[1])
        })
        .collect::<Vec<(i32, i32)>>()
}

fn extract(pairs: &Vec<(i32, i32)>, f: fn(&(i32, i32)) -> i32) -> Vec<i32> {
    let mut values: Vec<_> = pairs
        .into_iter()
        .map(f)
        .collect();
    values.sort();    
    values
}

fn part1(pairs: &Vec<(i32, i32)>) {
    let firsts: Vec<i32> = extract(&pairs, |p| p.0);
    let seconds: Vec<i32> = extract(&pairs, |p| p.1);

    let mut diff = 0;
    for i in 0..firsts.len() {
        diff += (firsts[i] - seconds[i]).abs();
    }    
    println!("Part 1: {:?}", diff);
}

fn frequencies(v: &Vec<i32>) -> HashMap<i32, i32> {
    v.iter()
        .fold(HashMap::new(), |mut map, val| {
              map.entry(*val)
                 .and_modify(|frq|*frq+=1)
                 .or_insert(1);
              map
          })
}

fn part2(pairs: &Vec<(i32, i32)>) {
    let firsts: Vec<i32> = extract(&pairs, |p| p.0);
    let seconds: Vec<i32> = extract(&pairs, |p| p.1);
    let freqs = frequencies(&seconds);

    let mut diff = 0;
    for el in firsts {
        diff += el * freqs.get(&el).unwrap_or(&0);
    }    
    println!("Part 2: {:?}", diff);
}

fn main() {
    let pairs = read_pairs("data/t01.txt");
    part1(&pairs);
    part2(&pairs);
}

File: clojure/2024/src/adv/t01.clj
(ns adv.t01
  (:require [adv.util :refer [split-lines split]]))

(defn parse-line [s]
  (->> (split s #"\s+")
       (map parse-long)))

(def input
  (->> "data/t01.txt"
       (slurp)
       (split-lines)
       (map parse-line)))

(defn distance [pairs]
  (let [ll (sort (map first pairs))
        rl (sort (map second pairs))]
    (->> (map #(Math/abs (- %1 %2))
              ll
              rl)
         (apply +))))

(defn part1 [] 
  (distance input))

(defn similarity-score [pairs]
  (let [freq1 (frequencies (map first pairs))
        freq2 (frequencies (map second pairs))]
    (->> freq1
         (map (fn [[key cnt]] (* key cnt (freq2 key 0))))
         (apply +))))

(defn part2 [] 
  (similarity-score input))
